docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,649 @@
+- `collab_type_name`: specifies that a server should only consider references to the given type name
+
+These parameters are emitted by the radicle p2p transport as headers in a 
+similar fashion to the `n=` parameter for the nonce or the `ls=true`
+parameter for selecting the `upload-pack-ls` service. The Radicle git server
+can then use these parameters to filter the refs it operates on.
+
+
+### Updating objects
+
+To make a change to an object we load the existing messages for an object. The 
+application developer provides us with the binary representation of the change
+to that object. We apply the change and ensure that the new object state still
+matches the object schema. At this point the state of the object may depend on
+many contributions from the tracking graph - not just the ones in our own view
+of the project. We now create a commit with our new change in it, referencing
+all the commits containing the direct dependencies of the change as parents.
+
+
+### Schema extensions
+
+#### `rad_signed_by`
+
+Many collaborative data structures will need to make statements about who is
+allowed to change what parts of a structure. To achieve this we extend the 
+json schema language with some custom metadata, the `rad_signed_by` property. This
+property can be placed on any `object` schema. It's value is an
+object with two keys, an array of properties which must be signed, and array of 
+radicle URNs who's signature must be present.
+
+This property implies a required `signatures` property which contains a mapping
+from URN to a description of the signature. (Maybe a JWS?).
+
+Any schema which has this property will result in some additional validation.
+Librad will encode the relevant keys of the target object using canonical JSON
+and then check that a signature over them is valid with respect to the given
+keys. 
+
+
+#### `frozen`
+
+Some attributes should never be changed, for example the ID of an issue, or a
+nonce on a comment. Any schema can add the metadata key, `frozen: true` to
+indicate that after it's initial creation any change which modifies it is
+invalid.
+
+#### `automerge_type`
+
+This can take the value `"text"` if  placed on `string` properties to indicate
+that they should be represented in an `Automerge.Text` data type.
+
+
+### Strange Perspectives
+
+This model introduces some counter-intuitive properties. For example, I might
+"create an issue" in a repository and anyone who is tracking me would see that
+issue, but people who are tracking the project but don't have me in their
+tracking graph will only see the issue if the maintainer replies to it. It's
+hard to see how you would do things like "link to an issue" under these
+constraints. This is inherent to the network model though, rather than being a
+specific problem of this architecture.
+
+We can work around some of this weirdness using seed nodes. If we consider
+seed nodes 
+
+## APIs
+
+The APIs librad will provide:
```

Comment:
I'm wondering if it's worth adding retrieval of schemas to the API? This will likely be needed for modifying them in the future proposal as well.

Reply:
The difficulty here is that in principle each object can have a different schema. This means that retrieving a schema is the same as retrieving an object (as the object includes its schema). It would be nice to require that all schemas of a type have the same schema - however, I think this introduces a griefing opportunity where someone publishes objects of a given type with a different schema to an already agreed schema and this poisons that type for anyone who syncs with them before syncing with well behaved remotes. I am still trying to figure out a nice way of handling this situation.

Reply:
> as the object includes its schema

Ah yes, that's a good point :) 

> I think this introduces a griefing opportunity where someone publishes objects of a given type with a different schema to an already agreed schema and this poisons that type for anyone who syncs with them before syncing with well behaved remotes.

Ya, I suppose this would fall under the umbrella of "consensus on schemas", where privileged persons can agree on the schema and a change would need validation from those persons.

Reply:
Yep. Now, one downside of not requiring that every object be of a particular schema is that the work of schema validation then falls on applications. This is not a deal breaker (as Mongodb has conclusively shown) but it's not ideal. One workaround might be to allow applications to specify a schema when they fetch objects, any object which doesn't match that schema would not be returned.

Reply:
Ya, to be honest, I prefer the idea of a zoo of compatible schemas. It means that while a project might have its own schema I could add additional metadata that I find interesting but which may be ignored by the project maintainers.

Reply:
But I guess the question then is: How does one _propose_ a change to the "agreed upon" schema?